apps/prairielearn/src/sprocs/ip_to_mode.sql

```diff
@@ -22,63 +20,66 @@ BEGIN
-                -- The reservation is in a testing center location that requires
-                -- network filtering.
-
-                PERFORM *
-                FROM pt_location_networks AS ln
-                WHERE
-                    ln.location_id = v_location_id
-                    AND ip <<= ln.network;
-
-                IF FOUND THEN
-                    -- The user is physically inside the testing center. Set
-                    -- mode to 'Exam'.
-                    mode := 'Exam';
-                ELSE
-                    -- Although we have a checked-in reservation, the user is
-                    -- physically outside the testing center. Set mode to
-                    -- 'Public'.
-                    mode := 'Public';
-                END IF;
-            ELSE
-                -- The reservation is in a testing center location that doesn't
-                -- require network filtering. Set mode to 'Exam'.
+            e.user_id = ip_to_mode.authn_user_id
+            AND (
+                (r.access_end IS NULL and s.date BETWEEN ip_to_mode.date - '1 hour'::interval and ip_to_mode.date + '1 hour'::interval)
+                OR (ip_to_mode.date BETWEEN r.access_start AND r.access_end)
+            )
+    LOOP
+        IF reservation.location_id IS NULL THEN
+            -- The reservation isn't for a testing center location (it's a
+            -- course-run session). If the reservation is "active", we're
+            -- in 'Exam' mode, and we return immediately. Otherwise, we might
+            -- be in 'Public' mode, but we continue looping to see if we have
+            -- any other reservations that might put us in 'Exam' mode.
+            IF reservation.reservation_active THEN
                 mode := 'Exam';
+                RETURN;
             END IF;
-        ELSE
-            -- The reservation isn't for a testing center location (it's a
-            -- course-run session), so we set exam mode.
+
+            mode := 'Public';
+            CONTINUE;
+        END IF;
+
+        IF NOT reservation.location_filter_networks THEN
+            -- The reservation is in a testing center location that doesn't
+            -- require network filtering. Set mode to 'Exam'.
             mode := 'Exam';
+            RETURN;
         END IF;
 
-    END IF;
+        PERFORM *
+        FROM pt_location_networks AS ln
+        WHERE
+            ln.location_id = reservation.location_id
+            AND ip <<= ln.network;
 
+        IF FOUND THEN
+            -- The user is physically inside the testing center. Set
+            -- mode to 'Exam'.
+            mode := 'Exam';
+            RETURN;
+        ELSE
+            -- Although we have a checked-in reservation, the user is
+            -- physically outside the testing center. Set mode to
+            -- 'Public', but continue looping to see if we have any other
+            -- reservations that might put us in 'Exam' mode.
+            mode := 'Public';
```

Comment:
Should we add a CONTINUE here to make it clear we should move on to consider the next reservation, like we did above for the earlier Public case? Of course it doesn't make any actual difference to the code execution.

Reply:
Yep, I can do that!

Reply:
Done in d0b33a4.